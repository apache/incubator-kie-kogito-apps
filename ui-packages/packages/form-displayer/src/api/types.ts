export interface FormDisplayerInitArgs {
  form: Form;
  data?: any;
  context?: Record<string, any>;
}

export type FormSubmitContext = {
  params?: Record<string, string>;
};

export enum FormOpenedState {
  OPENED = 'opened',
  ERROR = 'error'
}

export type FormOpened = {
  state: FormOpenedState;
  size: FormSize;
};

export type FormSize = {
  width: number;
  height: number;
};

export enum FormSubmitResponseType {
  SUCCESS = 'success',
  FAILURE = 'failure'
}

export type FormSubmitResponse = {
  type: FormSubmitResponseType;
  info: any;
};

export interface Association {
  origin: string;
  envelopeServerId: string;
}

export interface FormResources {
  scripts: {
    [key: string]: string;
  };
  styles: {
    [key: string]: string;
  };
}
interface FormConfiguration {
  schema: string;
  resources: FormResources;
}

export interface Form {
  formInfo: FormInfo;
  source: string;
  configuration: FormConfiguration;
}

export enum FormType {
  HTML = 'HTML',
  TSX = 'TSX'
}

export interface FormInfo {
  name: string;
  type: FormType;
  lastModified: Date;
}
