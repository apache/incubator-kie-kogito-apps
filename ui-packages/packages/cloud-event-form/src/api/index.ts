export * from './CloudEventFormApi';
export * from './CloudEventFormChannelApi';
export * from './CloudEventFormDriver';
export * from './CloudEventFormEnvelopeApi';
