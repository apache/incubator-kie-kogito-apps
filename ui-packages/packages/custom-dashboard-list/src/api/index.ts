export * from './CustomDashboardListApi';
export * from './CustomDashboardListChannelApi';
export * from './CustomDashboardListEnvelopeApi';
export * from './CustomDashboardListDriver';
