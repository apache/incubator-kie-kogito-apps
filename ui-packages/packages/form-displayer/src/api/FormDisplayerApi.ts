import {
  FormDisplayerInitArgs,
  FormSubmitContext,
  FormSubmitResponse
} from './types';

export interface FormDisplayerApi {
  init: (args: FormDisplayerInitArgs) => void;

  startSubmit: (context: FormSubmitContext) => Promise<any>;

  notifySubmitResult: (result: FormSubmitResponse) => void;
}
