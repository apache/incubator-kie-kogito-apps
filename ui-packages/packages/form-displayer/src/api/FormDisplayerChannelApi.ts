import { FormOpened } from './types';

export interface FormDisplayerChannelApi {
  notifyOnOpenForm: (opened: FormOpened) => void;
}
