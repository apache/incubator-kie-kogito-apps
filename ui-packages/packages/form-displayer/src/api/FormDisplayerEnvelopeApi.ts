import {
  Association,
  FormDisplayerInitArgs,
  FormSubmitContext,
  FormSubmitResponse
} from './types';

export interface FormDisplayerEnvelopeApi {
  formDisplayer__init(
    association: Association,
    initArgs: FormDisplayerInitArgs
  ): Promise<void>;
  formDisplayer__startSubmit(context: FormSubmitContext): Promise<any>;
  formDisplayer__notifySubmitResponse(result: FormSubmitResponse): void;
  formDisplayer__notifyInit(initArgs: FormDisplayerInitArgs): void;
}
