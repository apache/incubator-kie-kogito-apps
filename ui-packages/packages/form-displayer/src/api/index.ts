export * from './types';
export * from './FormDisplayerApi';
export * from './FormDisplayerChannelApi';
export * from './FormDisplayerEnvelopeApi';
