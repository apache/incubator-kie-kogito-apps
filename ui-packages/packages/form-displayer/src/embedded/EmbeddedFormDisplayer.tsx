/*
 * Copyright 2021 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { useCallback, useMemo } from 'react';
import {
  FormDisplayerApi,
  FormDisplayerChannelApi,
  FormDisplayerEnvelopeApi,
  Form,
  FormDisplayerInitArgs,
  FormOpened
} from '../api';
import { ContainerType } from '@kogito-tooling/envelope/dist/api';
import { EnvelopeServer } from '@kogito-tooling/envelope-bus/dist/channel';
import { EmbeddedEnvelopeFactory } from '@kogito-tooling/envelope/dist/embedded';

export type Props = {
  targetOrigin: string;
  formContent: Form;
  data?: any;
  context?: Record<string, any>;
  envelopePath: string;
  onOpenForm?: (opened: FormOpened) => void;
};

export const EmbeddedFormDisplayer = React.forwardRef<FormDisplayerApi, Props>(
  (props, forwardedRef) => {
    const pollInit = useCallback((
      // eslint-disable-next-line
      envelopeServer: EnvelopeServer<
        FormDisplayerChannelApi,
        FormDisplayerEnvelopeApi
      >
    ) => {
      return envelopeServer.envelopeApi.requests.formDisplayer__init(
        {
          origin: envelopeServer.origin,
          envelopeServerId: envelopeServer.id
        },
        {
          form: props.formContent,
          data: props.data ?? {},
          context: props.context ?? {}
        }
      );
    }, []);

    const refDelegate = useCallback(
      (
        envelopeServer: EnvelopeServer<
          FormDisplayerChannelApi,
          FormDisplayerEnvelopeApi
        >
      ): FormDisplayerApi => ({
        startSubmit: context => {
          return envelopeServer.envelopeApi.requests.formDisplayer__startSubmit(
            context
          );
        },
        notifySubmitResult: response => {
          envelopeServer.envelopeApi.notifications.formDisplayer__notifySubmitResponse(
            response
          );
        },
        init: (args: FormDisplayerInitArgs) => {
          envelopeServer.envelopeApi.notifications.formDisplayer__notifyInit(
            args
          );
        }
      }),
      []
    );

    const EmbeddedEnvelope = useMemo(() => {
      return EmbeddedEnvelopeFactory({
        api: {
          notifyOnOpenForm: opened => {
            if (props.onOpenForm) {
              props.onOpenForm(opened);
            }
          }
        },
        origin: props.targetOrigin,
        refDelegate,
        pollInit,
        config: {
          containerType: ContainerType.IFRAME,
          envelopePath: props.envelopePath
        }
      });
    }, []);

    return <EmbeddedEnvelope ref={forwardedRef} />;
  }
);
