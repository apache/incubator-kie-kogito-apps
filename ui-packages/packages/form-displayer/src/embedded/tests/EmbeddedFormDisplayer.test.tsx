/*
 * Copyright 2021 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { EmbeddedFormDisplayer } from '../EmbeddedFormDisplayer';
import { mount } from 'enzyme';
import { FormType } from '../../api';

describe('EmbeddedFormDisplayer tests', () => {
  it('Snapshot', () => {
    const props = {
      targetOrigin: 'origin',
      envelopePath: '/resources/form-displayer.html',
      formContent: {
        formInfo: {
          name: 'react_hiring_HRInterview',
          lastModified: new Date('2021-08-23T13:26:02.130Z'),
          type: FormType.TSX
        },
        configuration: {
          resources: {
            scripts: {},
            styles: {}
          },
          schema: 'json schema'
        },
        source: 'react source code'
      }
    };

    const wrapper = mount(<EmbeddedFormDisplayer {...props} />);

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.props().targetOrigin).toStrictEqual(props.targetOrigin);
    const contentIframe = wrapper.find('iframe');

    expect(contentIframe.exists()).toBeTruthy();
  });
});
