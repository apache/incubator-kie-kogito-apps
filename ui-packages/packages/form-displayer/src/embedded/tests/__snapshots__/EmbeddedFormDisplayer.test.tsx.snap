// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmbeddedFormDisplayer tests Snapshot 1`] = `
<ForwardRef
  envelopePath="/resources/form-displayer.html"
  formContent={
    Object {
      "configuration": Object {
        "resources": Object {
          "scripts": Object {},
          "styles": Object {},
        },
        "schema": "json schema",
      },
      "formInfo": Object {
        "lastModified": 2021-08-23T13:26:02.130Z,
        "name": "react_hiring_HRInterview",
        "type": "TSX",
      },
      "source": "react source code",
    }
  }
  targetOrigin="origin"
>
  <ForwardRef>
    <iframe
      src="/resources/form-displayer.html"
      style={
        Object {
          "border": "none",
          "display": "flex",
          "flex": 1,
          "flexDirection": "column",
          "height": "100%",
          "margin": 0,
          "overflow": "hidden",
          "padding": 0,
          "width": "100%",
        }
      }
      title="X"
    />
  </ForwardRef>
</ForwardRef>
`;
